### PR TITLE
feat: updates high priority functions to v9 API

### LINKED ISSUE #23: Migrate to Toggl API v9 and Reports API v3
Toggl has once again decided that they need to update their public API: https://developers.track.toggl.com/docs/index.html

> This documentation focuses on current API v9 and the old v8 version will soon be deprecated. The documentation is already extended to [Webhooks API](https://developers.track.toggl.com/docs/webhooks_start) and will soon be extended to missing services like Reports API v3.

It will probably be a while until the v8 and reports v2 APIs are deprecated, but this issue can serve as a discussion on the migration.

### PR DESCRIPTION
**Pull request by beauraines**
Updating to the new V9 API, focusing on the methods needed for [Obsidian integration for Toggl](https://github.com/mcndt/obsidian-toggl-integration) and [toggl-cli-node](https://github.com/beauraines/toggl-cli-node).  Other methods will be undated separately as a fast follow.

- [x] client.timeEntries.current (toggl-cli-node) 53148e2
- [x] client.timeEntries.list  (toggl-cli-node) cc483b
- [x] client.workspaces.list (common) e6ce389
- [x] client.workspaces.projects (common) No change needed https://developers.track.toggl.com/docs/api/workspaces#get-get-single-workspace
- [x] client.timeEntries.create  (toggl-cli-node)
- [x] client.timeEntries.stop (toggl-cli-node)
- [x] workspaces.tags (obsidian)
- [x] reports.details (obsidian)
- [x]  reports.summary (obsidian)
- [x] client.user.current (toggl-cli-node)
- [x] client.reports.weekly (toggl-cli-node)
- [x] client.timeEntries.update (toggl-cli-node)

There are many breaking changes in the v9 API
1. the reports and time entries list require `start_date` and/or `end_date` in the request.
2. Where `pid` and `wid` were used before, most of them switch to `project_id` and `workspace_id`
3. The API no longer returns the information in a `data` property, it basically moved it up one level.

Resolves #23

### CODE REVIEW

File: lib/client.js
Comment (beauraines) [inline lib/client.js:78]: This comes from PR #24

**Comment by beauraines** [comment]
When I resolved the merge conflicts, this brought in all the ESM imports rather than the old skool requires. I'm not sure how to _use_ the package once its been updated to ESM or really how to upgrade _my_ app to use the ESM version... But I'm about to call it a night.

**Comment by beauraines** [comment]
Moved to draft because I think that the version number needs to be incremented and probably the README, too. Plus I'm concerned about the ESM changes and what it means for my app...

**Comment by beauraines** [comment]
Tests failing due to missing secret it seems.

> Please make sure to set the environment variable "TOGGL_API_TOKEN" before running the smoke tests

**Comment by saintedlama** [comment]
> When I resolved the merge conflicts, this brought in all the ESM imports rather than the old skool requires. I'm not sure how to _use_ the package once its been updated to ESM or really how to upgrade _my_ app to use the ESM version... But I'm about to call it a night.

Those import dependencies already have been introduced with version 2.0 - see the CHANGELOG.md -https://github.com/saintedlama/toggl-client/blob/master/CHANGELOG.md

As long as the 2.0 version has been used already you'll not face any problems updating to 3.0 and beyond

**Comment by saintedlama** [comment]
> Tests failing due to missing secret it seems.
> 
> > Please make sure to set the environment variable "TOGGL_API_TOKEN" before running the smoke tests

The secret is configured in GitHub actions but is not accessible in the PR. I think we could use a "let's see when merging this to master approach" - issues still can be fixed then...

**Comment by saintedlama** [comment]
> Moved to draft because I think that the version number needs to be incremented and probably the README, too. Plus I'm concerned about the ESM changes and what it means for my app...

As this repo uses semantic commits - a "BREAKING CHANGE:" in the commit message body is enough to "hint" the major version increment. I can do it once merging the PR using the "squash and merge" strategy via GitHub

**Comment by saintedlama** [comment]
Sorry for causing the next merge conflict via package-lock.json - did release the first adjustments for using the v9 endpoint

**Comment by mcndt** [comment]
I will check out your patches and do some testing against my Obsidian client.

**Comment by beauraines** [comment]
LOL 😁 I cheated and made a non ESM fork that had your one or two changes. I hadn't been using 2.0 yet. I had struggled with converting my app to ESM. So, this is forcing me to do that. That said, I found [cjs-to-es6](https://www.npmjs.com/package/cjs-to-es6) and it was very helpful. I had to do some manual changes, but I'm making good headway and know what I need to do. I hope to finish that this weekend, too. I'm fighting with [yargs](http://yargs.js.org/).  If you have any recommendations for CLI parsing packages that work nicely with ESM6, I'd love to hear. Yargs did well with reading commands from a directory of JS files, which was why I chose it. That feature is borked with ESM6 unfortunately.

So this is good, its forcing me into the future!!

> > When I resolved the merge conflicts, this brought in all the ESM imports rather than the old skool requires. I'm not sure how to _use_ the package once its been updated to ESM or really how to upgrade _my_ app to use the ESM version... But I'm about to call it a night.
> 
> Those import dependencies already have been introduced with version 2.0 - see the CHANGELOG.md -https://github.com/saintedlama/toggl-client/blob/master/CHANGELOG.md
> 
> As long as the 2.0 version has been used already you'll not face any problems updating to 3.0 and beyond

**Comment by saintedlama** [comment]
> Tests failing due to missing secret it seems.
> 
> > Please make sure to set the environment variable "TOGGL_API_TOKEN" before running the smoke tests

You can configure a toggl api token in the forked repository under settings -> secrets & variables -> actions in "Repository secrets"

File: test/smoke-test.js
Comment (beauraines) [inline test/smoke-test.js:63]: @mcndt @saintedlama I'm not sure how this unit test is supposed to be written, but I'm trying to ensure that the exception handling is working

File: test/smoke-test.js
Comment (beauraines) [inline test/smoke-test.js:97]: Same with this one.

**Comment by beauraines** [comment]
OKAY!!! I think that this ready to be merged. I've successfully updated my CLI app to ESM6 and got it running with both v2 and this branch of toggl-client.

There are two unit tests that I need some help with and I would appreciate an extra eye on this, but I think that this is ready to be merged. As a new collaborator, I'm not sure if I should just merge it or not...

**Comment by saintedlama** [comment]
> @saintedlama
> 
> OKAY!!! I think that this ready to be merged. I've successfully updated my CLI app to ESM6 and got it running with both v2 and this branch of toggl-client.
> 
> There are two unit tests that I need some help with and I would appreciate an extra eye on this, but I think that this is ready to be merged. As a new collaborator, I'm not sure if I should just merge it or not...

I'm not sure either 😄 let me merge the PR to use the secrets to run the test, then I'll have a look at the two borked unit tests, and let me review the PR then. Then we're good to merge. I think that a review workflow can work. Btw: There is some functionality in the GitHub actions that when pushing or merging some commits to master a new version is published to npm automatically.

File: test/smoke-test.js
Comment (saintedlama) [inline test/smoke-test.js:61]: @beauraines fixed the tests by using good old plain try/catch blocks and failing in case no exception is thrown

**Review by saintedlama — APPROVED** [review]

**Comment by beauraines** [comment]
I see the strategy in how you updated those tests. I was thinking that the error would bubble up, rather than having to try and catch it in the smoke tests!

Auto-publishing is cool, I'm really going to have to checkout how you've setup these Github actions.

I really thought #31 would allow the unit tests to run on this PR... 🤷 

But I think we're there, and as you said, "let see what happens". 👍

**Comment by beauraines** [comment]
Github Actions ran!!! And we got rate limited!!! This seems like a good thing!

**Comment by saintedlama** [comment]
@beauraines wanna press the "Squash and merge" merge button?

**Comment by saintedlama** [comment]
> Github Actions ran!!! And we got rate limited!!! This seems like a good thing!

Removed the node 19 build from the build matrix as a quick fix AND it did run and publish to npm. 

🎉 congrats on the first release and the awesome functionality added through the PR!

**Comment by beauraines** [comment]
Thanks! I'm excited about this. I learned a bunch this week and I look
forward to collaborating with you.

--
Beau Raines
***@***.***
425.654.4663 (voice and sms)
beau.raines (skype)
LinkedIn www.linkedin.com/in/beauraines/

**Pull request merged**
